File: ode/kepler/octave/kepler.m
# Kepler problem

eps = 0.6
N = 120
K = 24    # Number of periods

q01 = 1 - eps;
q02 = 0;
u01 = 0;
u02 = sqrt((1 + eps)/(1 - eps));

function xdot = f(x,t)
  r = hypot(x(1), x(2));
  xdot(1) = x(3);
  xdot(2) = x(4);
  xdot(3) = -x(1)/r^3;
  xdot(4) = -x(2)/r^3;
endfunction

function e = energy(q1, q2, p1, p2)
  e = 0.5*(p1.^2 + p2.^2) - 1/hypot(q1, q2);
endfunction

function l = angular(q1, q2, p1, p2)
  l = q1*p2 - p1*q2;
endfunction

function configuration_plot(x, eps)
  global K;
  gap = 0.1;
  marker_size = 3.5;

  hf = figure();
  scatter(x(:, [1]), x(:, [2]), marker_size);
  axis equal
  axis([-1 - eps - gap, 1 - eps + gap, -sqrt(1 - eps^2) - gap, sqrt(1 - eps^2) + gap]);
  set(gca, 'fontsize', 11);
  grid on
  title "Kepler problem"
  xlabel "x"
  ylabel "y"

  glim = axis();
  text(
    glim(1) + 0.05*(glim(2) - glim(1)),
    glim(4) - 0.05*(glim(4) - glim(3)),
    sprintf("\\epsilon = %4.3f\nPeriods: %d", eps, K)
  );
  print(hf, "orbit.jpg", "-S960,720");
  print(hf, "orbit.eps", "-S960,720");
  #print(hf, "orbit.pdf", "-S960,720");
endfunction

function invariant_plot(x, t)
  marker_size = 3.5;

  e0 = energy(x(1, 1), x(1, 2), x(1, 3), x(1, 4));
  l0 = angular(x(1, 1), x(1, 2), x(1, 3), x(1, 4));
  m = 1;
  for el = x'
    sample(m).e = energy(el(1), el(2), el(3), el(4)) - e0;
    sample(m).l = angular(el(1), el(2), el(3), el(4)) - l0;
    m++;
  endfor

  hf = figure();
  scatter(t/(2*pi), [sample.e], marker_size);
  hold on
  scatter(t/(2*pi), [sample.l], marker_size, "marker", "+");
  set(gca, 'fontsize', 11);
  grid on
  title "Kepler problem"
  xlabel "Periods"
  ylabel "Error"
  legend("energy", "angular momentum");

  print(hf, "error.jpg", "-S960,720");
  print(hf, "error.eps", "-S960,720");
endfunction

x = lsode("f", [q01; q02; u01; u02], (t = linspace (0, 2*pi*K, N*K + 1)'));

configuration_plot(x, eps);
invariant_plot(x, t);

File: ode/kepler/python/kepler.py
import numpy as np
import matplotlib.pyplot as plt

from scipy.integrate import solve_ivp

eps = 0.6
N_ppr = 120   # Number of points per rotation
N_rot = 25    # Number of rotations
t_step = 2*np.pi/N_ppr
x_init = [1 - eps, 0, 0, np.sqrt((1 + eps)/(1 - eps))]

def dfunc(t, x):
    res = np.zeros_like(x)
    r = np.hypot(x[0], x[1])
    res[0] = x[2]
    res[1] = x[3]
    res[2] = -x[0]/r**3
    res[3] = -x[1]/r**3
    return res

def energy(x):
    E = 0.5*np.hypot(x[2], x[3])**2 - 1/np.hypot(x[0], x[1])
    return E

def angular(x):
    L = x[0]*x[3] - x[2]*x[1]
    return L

sol = solve_ivp(dfunc, [0, N_rot*2*np.pi], x_init, t_eval=[n*t_step for n in range(0, N_rot*N_ppr + 1)], method="DOP853")

fig = plt.figure(figsize=(8, 8))
plt.title("Kepler problem. Orbit")
plt.axis("equal")
plt.xlabel("x")
plt.ylabel("y")
plt.grid(True)

#plt.scatter(sol.y[0], sol.y[1])
plt.plot(sol.y[0], sol.y[1])
xmin, xmax, ymin, ymax = plt.axis()
plt.text(xmin + 0.05*(xmax - xmin), ymax - 0.03*(xmax - xmin), "eps = {}".format(eps))
fig.savefig('orbit.png', dpi=150)

e0 = energy(x_init)
l0 = angular(x_init)
fig = plt.figure(figsize=(8, 8))
plt.title("Kepler problem. Invariant error")
plt.xlabel("Periods")
plt.ylabel("Error")
plt.grid(True)

plt.plot(sol.t/(2*np.pi), energy(sol.y) - e0)
plt.plot(sol.t/(2*np.pi), angular(sol.y) - l0)
plt.legend(['Energy', 'Angular momentum'])
fig.savefig('error.png', dpi=150)
